### PR TITLE
WIP : Move files and directories in a workspace

### PR DESCRIPTION
**Pull request by c4ffein**

### CODE REVIEW

**Comment by c4ffein** [comment]
Resolves #435

File: parsec/core/fs/workspacefs/workspacefs.py [hunk 1]
```diff
@@ -186,12 +186,55 @@ def workspace_name(self) -> str:
...
+        # If real_destination is not found, we can continue
+        except FileNotFoundError as e:
+            pass
+        if await self.is_dir(source):
+            await self.copytree(source, real_destination)
+            await self.rmtree(source)
+            return
+        elif await self.is_file(source):
+            await self.copyfile(source, real_destination)
+            await self.entry_transactions.file_delete(source)
```
Comment (vxgmichel) [inline parsec/core/fs/workspacefs/workspacefs.py]: Maybe avoid using low level transactions here? `self.unlink(source)` should work fine

File: parsec/core/fs/workspacefs/workspacefs.py [hunk 2]
```diff
@@ -186,12 +186,55 @@ def workspace_name(self) -> str:
...
         destination = FsPath(destination)
+        real_destination = destination
         if source.parent == destination.parent:
             return await self.rename(source, destination)
-        # TODO - reference implementation:
-        # https://github.com/python/cpython/blob/3.7/Lib/shutil.py#L525
+        try:
+            if await self.is_dir(destination):
+                real_destination = destination.joinpath(source.name)
+                if await self.path_info(real_destination):
```
Comment (vxgmichel) [inline parsec/core/fs/workspacefs/workspacefs.py]: This make me realize I forgot to add the [exists](https://docs.python.org/3.5/library/pathlib.html#pathlib.Path.exists) method, feel free to add it :)

File: parsec/core/fs/workspacefs/workspacefs.py [hunk 3]
```diff
@@ -186,12 +186,55 @@ def workspace_name(self) -> str:
...
+                    raise FileExistsError
+        # If real_destination is not found, we can continue
+        except FileNotFoundError as e:
```
Comment (vxgmichel) [inline parsec/core/fs/workspacefs/workspacefs.py]: This `try/except` is a bit wide, could we shrink it?

File: parsec/core/fs/workspacefs/workspacefs.py [hunk 4]
```diff
@@ -186,12 +186,55 @@ def workspace_name(self) -> str:
...
+            await self.copytree(source, real_destination)
+            await self.rmtree(source)
+            return
+        elif await self.is_file(source):
+            await self.copyfile(source, real_destination)
+            await self.entry_transactions.file_delete(source)
+            return
         raise NotImplementedError
 
+    def _destinsrc(src: AnyPath, dst: AnyPath):
```
Comment (vxgmichel) [inline parsec/core/fs/workspacefs/workspacefs.py]: A more explicit name maybe? `_is_destination_in_source` ?

File: parsec/core/fs/workspacefs/workspacefs.py [hunk 5]
```diff
@@ -186,12 +186,55 @@ def workspace_name(self) -> str:
...
+        try:
+            dst.relative_to(src)
+            return True
+        except ValueError:
+            return False
+
+    async def copytree(self, source_path: AnyPath, target_path: AnyPath):
+        source_files = await self.listdir(source_path)
+        await self.mkdir(target_path)
+        errors = []
```
Comment (vxgmichel) [inline parsec/core/fs/workspacefs/workspacefs.py]: This list seems to be unused

File: parsec/core/fs/workspacefs/workspacefs.py [hunk 6]
```diff
@@ -186,12 +186,55 @@ def workspace_name(self) -> str:
...
+        for source_file in source_files:
+            target_file = target_path.joinpath(source_file.name)
+            if await self.is_dir(source_file):
+                await self.copytree(source_file, target_file)
+            elif await self.is_file(source_file):
+                await self.copyfile(source_file, target_file)
+
+    async def copyfile(self, source_path: AnyPath, target_path: AnyPath, length=16 * 1024):
+        await self.touch(target_path)
+        off = 0
```
Comment (vxgmichel) [inline parsec/core/fs/workspacefs/workspacefs.py]: `offset` maybe?

File: parsec/core/fs/workspacefs/workspacefs.py [hunk 7]
```diff
@@ -186,12 +186,55 @@ def workspace_name(self) -> str:
...
+        while 1:
```
Comment (vxgmichel) [inline parsec/core/fs/workspacefs/workspacefs.py]: `while True`?

File: parsec/core/types/base.py [hunk 8]
```diff
@@ -64,13 +64,13 @@ def is_root(self):
...
+    # TODO: this method should be removed with the FS class
+    # The concept of workspace path doesn't apply anymore
+
     @property
     def workspace(self):
         return self.parts[1]
 
-    # TODO: this method should be removed with the FS class
-    # The concept of workspace path doesn't apply anymore
-
```
Comment (vxgmichel) [inline parsec/core/types/base.py]: Thanks for moving this!

File: parsec/core/fs/workspacefs/workspacefs.py (hunk 1 above)
Comment (c4ffein) [reply parsec/core/fs/workspacefs/workspacefs.py]: Done!

File: parsec/core/fs/workspacefs/workspacefs.py (hunk 2 above)
Comment (c4ffein) [reply parsec/core/fs/workspacefs/workspacefs.py]: Done!

File: parsec/core/fs/workspacefs/workspacefs.py (hunk 4 above)
Comment (c4ffein) [reply parsec/core/fs/workspacefs/workspacefs.py]: I followed the shutil naming, but as it is an internal method I can change it if you prefer

File: parsec/core/fs/workspacefs/workspacefs.py (hunk 5 above)
Comment (c4ffein) [reply parsec/core/fs/workspacefs/workspacefs.py]: Yes, I forgot it, it was part of my first implementation where I kept a list of failed copies and raised an error only at the end, mimicking shutil behaviour. Do you want me to re-implement it that way?

File: parsec/core/fs/workspacefs/workspacefs.py (hunk 6 above)
Comment (c4ffein) [reply parsec/core/fs/workspacefs/workspacefs.py]: Changed it

File: parsec/core/fs/workspacefs/workspacefs.py (hunk 7 above)
Comment (c4ffein) [reply parsec/core/fs/workspacefs/workspacefs.py]: changed it, old habit from python 2 where it was better optimised

File: parsec/core/fs/workspacefs/workspacefs.py (hunk 3 above)
Comment (c4ffein) [reply parsec/core/fs/workspacefs/workspacefs.py]: It responds to the need of having real_destination referencing either the destination path if it didn't already exist, or a new entry with the same name as source, but inside the destination directory. I updated the comment, is it clear enough or should I try to implement it another way?

**Pull request merged**